Lambda as UDA is valid but spec requires brackets

## Richard (Rikki) Andrew Cattermole reported this on 2024-01-20T01:25:08Z

### Transferred from https://issues.dlang.org/show_bug.cgi?id=24346

### CC List

* 

### Description

````markdown
From: https://forum.dlang.org/post/hnjcuvuzytbuvgfdhfkd@forum.dlang.org

The lambda here is valid, even though it is not in a brackets.

```d
@a=>7 int b;
pragma(msg, __traits(getAttributes, b)); // AliasSeq!((a) => 7)
```

Should only be supported (according to specification):

```d
@(a=>7) int b;
pragma(msg, __traits(getAttributes, b)); // AliasSeq!((a) => 7)
```

Clarification or a compiler bug fix is needed.
````

Allowing this breaks shortened function bodies:

```d
struct UDA {}

int foo() @UDA => 123;

void main()
{
    assert(foo() == 123);
    // error: undefined reference to '_D9onlineapp3fooFZi'
}
```

In the above code, `=> 123` should be parsed as the function body of `foo`, but it is instead parsed as part of the attribute `@(UDA => 123)`.

--- a/compiler/src/dmd/parse.d
+++ b/compiler/src/dmd/parse.d
@@ -1319,7 +1319,21 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
 
             // Allow identifier, template instantiation, or function call
             // for `@Argument` (single UDA) form.
-            AST.Expression exp = parsePrimaryExp();
+            AST.Expression exp;
+
+            {
+                const loc = token.loc;
+                Identifier id = token.ident;
+                nextToken();
+                if (token.value == TOK.not)
+                {
+                    auto tempinst = new AST.TemplateInstance(loc, id, parseTemplateArguments());
+                    exp = new AST.ScopeExp(loc, tempinst);
+                }
+                else
+                    exp = new AST.IdentifierExp(loc, id);
+            }
+
             if (token.value == TOK.leftParenthesis)
             {
                 const loc = token.loc;

--- a/compiler/test/compilable/uda_lambda.d
+++ b/compiler/test/compilable/uda_lambda.d
@@ -1,0 +1,3 @@
+enum UDA;
+int fun() @UDA => 7;
+static assert(fun() == 7);

--- a/compiler/test/fail_compilation/uda_lambda.d
+++ b/compiler/test/fail_compilation/uda_lambda.d
@@ -1,0 +1,7 @@
+/+
+TEST_OUTPUT:
+---
+fail_compilation/uda_lambda.d(7): Error: declaration expected, not `=>`
+---
++/
+@a => 7 int b;

